tests/utils.bash

```diff
@@ -0,0 +1,52 @@
+report_lines ()
+{
+    local line
+
+    if [[ ${#lines[@]} -gt 0 ]]; then
+	echo "# "
+    fi
+    
+    for line in "${lines[@]}"; do
+	echo "# $line"
+    done
+} >&3
+
+#
+# simple_test input.peg
+# 
+simple_test ()
+{
+    local cmd
+
+    local input=$1
+    local base=$(basename "$input" .peg)
+    local generated_base=tests/"$base".generated
+    local expected_base=tests/"$base".expected
+
+    # force inserting new line to the output
+    echo "# " >&3
+    for cmd in build/*/bin/*/packcc; do
+	if ! [[ -x "$cmd" ]]; then
+	    continue
+	fi
+
+	echo "# target executable: $cmd" >&3
+	
+	# Check exist status
+	run "$cmd" -o "$generated_base" "$input"
+	[ "$status" -eq 0 ]
+
+	# Compare .c file
+	run diff -uN "$expected_base".c "$generated_base".c
```

Comment:
Thank you for reviewing.

> I'm afraid that this is a very bad idea. Comparing entire generate code for each test will make the tests very hard to maintain, because any change to generated code will break all the tests. Even slightest change in whitespace or wording of comment, which doesn't have any significance will break all tests, giving user no useful information about what's wrong.

I don't think so. I think any code changes that change generated code should be tested.
Quoted from README.md
```
Some additional features are as follows:

    Thread-safe and reentrant,
    Generates more ease-of-understanding parser source codes,
```
I think whitespaces in the generated code take an important role in realizing "ease-of-understanding parser source codes".
@arithy, how do you think?

For the rest of your comments, give me more time to reflect on my pull request. 3 ~ 5 are quite important suggestions.

Reply:
Sorry, I probably didn't express it clear enough.

> I don't think so. I think any code changes that change generated code should be tested.

I do agree that it is important to test the generated code. But it must be in a meaningful and sustainable way. The way you proposed would mean that if you need to change the generated code (e.g. to fix a bug), you'll have to change all the expected files, because all of them would change. If there is many tests, then it will not be possible to do it manually and it will be error-prone. 

Therefore, I suggest to make each test check only those parts of the file, that contain the tested feature. E.g: Test for matching of character classes should only look at the code that deals with the character classes. It should not fail if someone changes string matching code or even when the PackCC version is increased (it is in the comment at the top of the generated files).Of course, there should be many tests, each checking some parts of the code, so together they check everything (or as much as possible). But if developer changes some code, it should only break tests related to that code, not everything.

Reply:
I would like to withdraw this pull request. I would like to create a new pull request focusing on item 5.